File: matrices_generator.py
import os
import numpy as np
import time

def generate_and_save_matrices(matrix_size_A, matrix_size_B, file_path_A, file_path_B):
    # Create directory if it doesn't exist
    os.makedirs(os.path.dirname(file_path_A), exist_ok=True)
    os.makedirs(os.path.dirname(file_path_B), exist_ok=True)

    # Generate random matrices
    matrix_A = np.random.randint(10, size=(matrix_size_A, matrix_size_B))
    matrix_B = np.random.randint(10, size=(matrix_size_B, matrix_size_A))

    # Save matrices to file
    with open(file_path_A, 'w') as file_A:
        for row in matrix_A:
            file_A.write(' '.join(map(str, row)) + '\n')

    with open(file_path_B, 'w') as file_B:
        for row in matrix_B:
            file_B.write(' '.join(map(str, row)) + '\n')


def multiply_matrices(file_path_A, file_path_B, result_file_path):
    # Load matrices from file
    with open(file_path_A, 'r') as file_A:
        matrix_A = [list(map(int, line.split())) for line in file_A]

    with open(file_path_B, 'r') as file_B:
        matrix_B = [list(map(int, line.split())) for line in file_B]

    if len(matrix_A[0]) != len(matrix_B):
        print("Matrices are not multiplicatively compatible.")
        return

    # Perform matrix multiplication with numpy
    start_time = time.time()
    result = np.dot(matrix_A, matrix_B)
    end_time = time.time()
    multiplication_time = end_time - start_time

    # Save result matrix to file
    with open(result_file_path, 'w') as result_file:
        for row in result:
            result_file.write(' '.join(map(str, row)) + '\n')

    # Save multiplication time to file
    report_file_path = result_file_path.split('.')[0] + "_report.txt"
    with open(report_file_path, 'w') as report_file:
        report_file.write("matrix multiplication using numpy took: {} seconds\n".format(multiplication_time))

matrix_size_A = 300  # Number of rows in matrix A
matrix_size_B = 400  # Number of columns in matrix A and rows in matrix B
file_path_A = "matrices/matrix_A.txt"
file_path_B = "matrices/matrix_B.txt"
result_file_path = "matrices/result.txt"

generate_and_save_matrices(
    matrix_size_A, matrix_size_B, file_path_A, file_path_B)

multiply_matrices(file_path_A, file_path_B, result_file_path)

File: cpp/src/matrices_generator.cpp
#include "matrices_generator.h"
#include <random>
#include <fstream>
#include <vector>
#include <chrono>
#include <numeric>
#include "matrix_operations.h"

void generate_and_save_matrices(int matrix_size_A, int matrix_size_B, const std::string& file_path_A, const std::string& file_path_B) {
    std::mt19937 rng(std::random_device{}());
    std::uniform_int_distribution<int> dist(0, 9);

    std::ofstream file_A(file_path_A);
    std::ofstream file_B(file_path_B);

    for (int i = 0; i < matrix_size_A; ++i) {
        for (int j = 0; j < matrix_size_B; ++j) {
            file_A << dist(rng) << " ";
        }
        file_A << "\n";
    }

    for (int i = 0; i < matrix_size_B; ++i) {
        for (int j = 0; j < matrix_size_A; ++j) {
            file_B << dist(rng) << " ";
        }
        file_B << "\n";
    }
}

double multiply_matrices(const std::string& file_path_A, const std::string& file_path_B, const std::string& result_file_path) {
    auto A = load_matrix_from_file(file_path_A);
    auto B = load_matrix_from_file(file_path_B);

    auto start_time = std::chrono::high_resolution_clock::now();
    auto result = single_thread_matrix_multiply(A, B);
    auto end_time = std::chrono::high_resolution_clock::now();
    double multiplication_time = std::chrono::duration<double>(end_time - start_time).count();

    std::ofstream result_file(result_file_path);
    for (const auto& row : result) {
        for (const auto& val : row) {
            result_file << val << " ";
        }
        result_file << "\n";
    }

    std::ofstream report_file(result_file_path + "_report.txt", std::ios_base::app);
    report_file << "Matrix multiplication using numpy took: " << multiplication_time << " seconds\n";

    return multiplication_time;
}
